### DISCUSSION TITLE
How can I refactor all the beanutils in my project to use mapstruct instead?

### DISCUSSION THREAD

**Discussion by dreamloserlvv** [mapstruct/mapstruct#3345] [General]
is there has any tools that can scan my project and replace all the beanutils to mapstruct

**Comment by filiphr** [comment]
I am not aware of any special tools that do this automatically. I know about [OpenRewrite](https://docs.openrewrite.org/) that you might be able to leverage to perform a migration. Might be worth a try

**Comment by Zongatel** [comment]
Hi,
I am on the same path. We are using PropertyUtils.copyProperties(...) and would like to get rid of that. We are using it within routines with generic types. Since MapStruct's working strategy is generating code, is there a good way to use MapStruct even with generic types? (In my imagination that is probably not possible, but I like to have a second opinion or may be a good hint or even a solution if possible.)

**Reply by filiphr** [reply]
I'll need to see some examples @Zongatel.

However, if you know all your types you can define mappers for them and you can even have something where you store mappings between different types (in a generic way) and then use that to get the right mapper and invoke the mapping method.

**Reply by Zongatel** [reply]
is it like this, that you need to test with instanceof the Type of the Generic within your generic routine and use the right fuction of the defined interface for MapStruct, or does it MapStruct with some magic in its generated code itself? Or in other words can you define in the interface also method-signatures with generics in and MapStruct does all the rest?

**Reply by filiphr** [reply]
For MapStruct you have to define the functions your self. However, you can write your own annotation processor that is going to collect all the available mapping methods and then use that to lookup the right mapping method.
